### ISSUE TITLE
Regression in alternative_long_member_definitions behaviour, or docs miss?

### ISSUE THREAD

**Issue by rehael** [fsprojects/fantomas#1666]
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAwgBQCGAaogE6YAysAdgOYAqOiaAguVbYy7kwAti6AJqwArgDlEieE1gBVAM6IANAB0K6uhF0RiAZURw68LjXrNWHczyv8hoidNnzlazdr27iggJYUZpQWvNZgnMF2fIiCwmJSMnKKKgCUYAC8njq60AA2xEpKWXqIppCq4FAq6Oh%2BjEqoYADalRCg3lAAJFi4jZB16OWtupCdAG7EueLIaC0dEAAMFfNQAJKmpegGfgBeyMvzACzDEAC6wwC%2BB20nXT0zUANDHaMTUw9z8wCM196QALLEAAe1DqiGopQY6AEz2%2BACYFrdznorsN2n9uqw%2BgAjWCwXKwvSvSbTRqfDpw35EowAWz8cFy9HY6AAoqYAPIAM1BdH2twgnMmKiRl1%2B6KJmN6aEguPxhJG4xJH35YAAzFSRgZsMRoIgAEKITmwCiIAAKxAoxBpiHQlHlenQFGmIpRYtuo3uOLxBI1xPeZJVhw1EEgWp1%2BsNxvBsAA7pRoAVEOsxrAE7V6PbdI7nR1kbpUXpxQrPdLZT73Yr-bMVQBWYNQMO6g1Gk0KbC4CgJlTJ1PCPwZ%2BuC3LC3OitEVktQMuZv2k6srABs9dD2qbkZN%2BHyhXw9CU2eg6GNmYFQsQLvzbpekoeMu9M8rc%2BaKoA7MvGxGW4h-ogadi7YPT3PCAC10IsQ2vL05V9B9lRWAAON9Vw-KMd0ZOhjzAIcR28PNgMvDFJ1vKCKzeR9yW8ABORDw3YTlbQoHcaRpYgMOzM9R1dccr0I6doNI2DviWFUV3DZso1pel8QHFUsPYnCx0LCcsVLO8%2BKVAMVi%2BH5hPfWj6IkhlpJWNigLAECbm45Sp1Ukj1PneFl3WJA6HQdk6CYKgAHU-GhDDZNM8ywDAu4rKI8sr34jTvnVHSkPYCgJFMAARRBcj8Ol6NYp05L0XCzPwiVCKeNSqyfTSg2EwEgVWTlBFKFlh0QAwBGNdBvPgXz60OREOIvLiCNC4rbNK8i9C%2BOtKuBdZOT8IF2Q7YRjRZIFsBNQp%2B3Q%2Bsax6%2BTOMUyypUeFz70i%2ByOi%2BJdJqBAAlYxjXgdrOsDHbcoU0ClMO-pjpKsiVS%2BV8rtuuBAkkcRf0oLlVltGkGnrL4AoK4tQppcRclqNLeQAMWNZiakoaITrssrvgQ4Sgfu-5UfRsFsYoXHMuXaAhEtA9KAAfRjPwOphBH%2BsKwbvuG37NKoq74stTB2QoUE90emEupe3Q8sC4KPQFwYfoEikhJWAFgXF4hJelvw91B8GKEh6HYb%2B3n9oGz6UbRvwMcQWn6fxrFNaiiltN1g2jZl9BKadl23eEBnhKZi0dXojmud8233oOm8hoionRt0OFKSu0glT1Op4DqBg5Yw7rE4s%2B2U8FtORpVOEYt1qrMfEOgDw2-PTCLkuFfLoKPqrjWha17w4QqxvgSS2B0AAcRtZbVsQdb6G72tFbON6K-5z7eKH72R4mxuqedsE9UZaAAGs9RZ8%2BbSUNyDCtRBULBzaZMA3q8L5pHt5smvhfmOEl1dbSBjC7A06A4ylGiClGadAfIbXYKYb85trYrH8h-fKX9wI8V-hiU6xMKQA11gAaRkNgGqdU6DrAftaHkfI0Hv12n1O2W8B6E1risOEpNx5AganSJQAgV4rDLhglW-dILhTwenOuotdbbEYLkRA8UGBg02J5RA2J-gIHofMdBTDP4sO-jeHef9h56FVDreYkB2BpQYHQZurd0x0G2HY4Q4gTTyCcpsPsRldGMNentJOlcJHsP-h0VUvsrE2PonQPsYxoyMGQX%2BCgMC6jwN3H5fxSsN592TiEr2Z1vCqmzofJ2dDqBWmxPAYgm5YBKCLiAl2mTGq91VhBFSxFTF73MQ3KxSUTbEGxIovhJsBAGEwC5YEzTsIBOYUE1hfRSjwC5HQgwWBFGhLMboVUY8rFsmWdyMEGEYAUFyJySArTxEdMkRKfBGcICqgPlY0hiByEbBcusK%2BxBW7yzfi00RiNsGhRMVIjh8xVRAKsafb5586FKHioleA0g9yyGDtTXk881r1IyfWEyAKsEhR-p00FYSilEKsWsig9Ig7aOmTlbJKJKinAOP0JQmN6mNH8hcIAA)

#### Code

```fsharp
type C(aVeryLongType: AVeryLongTypeThatYouNeedToUse,
       aSecondVeryLongType: AVeryLongTypeThatYouNeedToUse,
       aThirdVeryLongType: AVeryLongTypeThatYouNeedToUse) =
    class
    end
```
            

#### Result

```fsharp
type C
    (
        aVeryLongType: AVeryLongTypeThatYouNeedToUse,
        aSecondVeryLongType: AVeryLongTypeThatYouNeedToUse,
        aThirdVeryLongType: AVeryLongTypeThatYouNeedToUse
    ) =
    class
    end

```

#### Problem description

Regression in [fsharp_alternative_long_member_definitions](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#fsharp_alternative_long_member_definitions) – expected no change whatsoever, got alternate formatting. Flipping the setting doesn't change the parenthesis placement formatting.

Is it caused by #1334 missing updating the docs?

Anyway, extra question then – why tuples are not taken into consideration with `MultilineBlockBracketsOnSameColumn`?

#### Options

Fantomas Master at 04/17/2021 11:23:36 - 26674f958aaa77992dcc5fdfc1bd43ce0d71e919

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

**Comment by nojaf** [comment]
Hello, thank you for showing interest in this project.
The documentation is not up to date so it seems.
In short, Fantomas follows two style guides: the one from [Microsoft](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#place-parameters-on-a-new-line-for-long-definitions) by default and the one from [G-Research](https://github.com/G-Research/fsharp-formatting-conventions#place-parameters-on-a-new-line-for-very-long-member-definitions) via various settings.
These style guides evolve over time so the delta of the `fsharp_alternative_long_member_definitions` setting became smaller as the Microsoft guide was updated.

Are you interested in submitting a PR to update the documentation?

> Anyway, extra question then – why tuples are not taken into consideration with MultilineBlockBracketsOnSameColumn?

MultilineBlockBracketsOnSameColumn was introduced to facilitate the [G-Research style](https://github.com/G-Research/fsharp-formatting-conventions#formatting-records) guide.

**Comment by rehael** [comment]
Sure, I can prepare a PR, maybe later today I will find some time. But first I need same clarity what should be the proper thing to do.

It looks like now the only difference this setting makes, is the placement of return type and `=` sign. Although I'm, not even sure about that, because G-Research style doesn't say anything explicitly about it. It looks like asking both style guides' authors what is their approach would have merit – because at then end of the day both teams might come to the same guidance. And then having this setting wouldn't have sense whatsoever.

What do you think?

**Comment by nojaf** [comment]
Well, there is a bit of history and context to this. Originally the Microsoft guide had a different view on things and I believe it was changed in https://github.com/dotnet/docs/pull/21690. The change was introduced to avoid "vanity alignments".

These things were discussed without having a clear view on the implementation side of things in Fantomas.
The style guide is not tied to Fantomas, nor was anybody involved from the G-Research style guide.
The conversation just kinda happened and they landed very close to what G-Research had all along.
So, this reflected in the setting being reduced to determining the location of the `=` sign.

Next, there is also something worth mentioning about the style guides. The Microsoft one is an open one. You can discuss and contribute to it via GitHub (there is a feedback button down below the page). The G-Research one is closed. They have published their guide to GitHub but it represents their vision.
So the only option, in this case, would be to pitch the G-Research style as a GitHub issue in the Microsoft one.
Perhaps it would fly, perhaps not. Ideally, it does and there is no need for a setting in Fantomas.

In case things stay as they are, the only thing that needs a change is the documentation in Fantomas.

To summarize, it sorta depends on how invested you are in all of this. You can start conversations on both sides, try and get the intent of why things are why they are and see how you can reconcile them.

**Comment by rehael** [comment]
For now, I created a PR with a docs fix.

Thank you for providing additional context! As for discussing further aligning the code style for constructors and member definitions with G-Research and Microsoft, then unless you would like to engage with them directly, I might open some GitHub issues to align that setting later this week.

**Comment by nojaf** [comment]
Feel free to engage with both guides yourself. Thank you for your efforts!
